### PR TITLE
Add configuration for Read the Docs build and pin docs requirements

### PR DESCRIPTION
**Pull request by samdoran**
Due to a recent release of `docutils` 0.18, our doc build is broken.
Add configuration for Read the Docs and pip docs requirements to ensure builds will not spontaneously break in the future.
Update tox configuration to match the build done by Read the Docs.

### CODE REVIEW

**Review by Shrews — APPROVED** [review]

File: .readthedocs.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+version: 2
```
Comment (webknjaz) [inline .readthedocs.yaml:1]: Note that the v2 config overrides all the web UI settings (unlike v1 that used to be merged).

Also, have you considered `htmldir` builder for nicer URLs? Plus maybe an option to be strict on warnings?

File: docs/requirements.txt [hunk 2]
```diff
@@ -1,3 +1,30 @@
-pbr
-sphinx
-ansible-core
+alabaster==0.7.12
```
Comment (webknjaz) [inline docs/requirements.txt:1]: Have you considered using pip-tools with `--generate-hashes`?

File: docs/requirements.in [hunk 3]
```diff
@@ -0,0 +1,3 @@
+pbr
+sphinx
+ansible-core
```
Comment (webknjaz) [inline docs/requirements.in]: Nit: maybe reorder this?

File: docs/requirements.txt (hunk 2 above)
Comment (samdoran) [reply docs/requirements.txt:1]: I have tried that in the past and ended up fighting it more than it was helping.

File: docs/requirements.in (hunk 3 above)
Comment (samdoran) [reply docs/requirements.in]: Doh! Can't believe I missed that. Thanks.

File: .readthedocs.yaml (hunk 1 above)
Comment (samdoran) [reply .readthedocs.yaml:1]: I looked at the web UI setting and didn't see anything worth preserving.

I didn't know about `htmldir`. I'll check it out.

I did think about being strict on warnings but noticed there was a warning in the build output I wasn't sure about how to fix it (missing `pexpect` I think?). Now that I have the build working I can try being strict about warnings.

Comment (samdoran) [reply .readthedocs.yaml:1]: I figured out the warning I was seeing and fixed it.

`dirhtml` is pretty cool. I think that might break existing links to documentation, though.

**Review by shanemcd — APPROVED** [review]

File: .readthedocs.yaml (hunk 1 above)
Comment (webknjaz) [reply .readthedocs.yaml:1]: @samdoran there's a built-in option in the RTD UI specifically to solve this problem. I described how to enable the exact setting you need here https://github.com/pyca/cryptography/issues/5863#issue-815633056.

File: .readthedocs.yaml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  - epub
+  - pdf
+
+build:
+  os: ubuntu-20.04
+  tools:
+    python: '3.10'
+
+sphinx:
+  builder: dirhtml
```
Comment (webknjaz) [inline .readthedocs.yaml:13]: The other conversation is marked as resolved so I'll link it here: https://github.com/ansible/ansible-runner/pull/888/files#r738479800.

TL;DR enable `Sphinx HTML -> HTMLDir` in the `Redirect type:` dropdown at https://readthedocs.org/dashboard/ansible-runner/redirects/.

File: tox.ini [hunk 5]
```diff
@@ -27,5 +27,6 @@ commands=
 [testenv:docs]
 description = Build documentation
 deps = -r{toxinidir}/docs/requirements.txt
+skip_install = True
 commands =
-    sphinx-build -E -W -d docs/build/doctrees -b html docs docs/build/html
+    sphinx-build -T -E -W --keep-going {tty:--color} -d docs/build/doctrees -b dirhtml docs docs/build/html
```
Comment (webknjaz) [inline tox.ini]: @samdoran I'd recommend keeping the `html` builder here because then, it'd be possible to view the docs locally w/o having to spawn a local web server, just via `file://` URLs.

But if you think this needs to match RTD, at least add a hint on how to browse the built docs. Here's what I typically add:
https://github.com/ansible/pylibssh/blob/a74c798/tox.ini#L301-L323

Comment (samdoran) [reply tox.ini]: Oh, that's weird. I figured a browser would just load the `index.html` in a folder. Evidently not.

Comment (webknjaz) [reply tox.ini]: No, that is typically supported by the web-server. Which is quite easy to spawn with `python -m http.server` and view via `http://localhost:...` but still some people may not like having to do so.

File: .readthedocs.yaml (hunk 4 above)
Comment (samdoran) [reply .readthedocs.yaml:13]: Sweet! I enabled that.

File: tox.ini [hunk 6]
```diff
@@ -27,5 +27,6 @@ commands=
 [testenv:docs]
 description = Build documentation
 deps = -r{toxinidir}/docs/requirements.txt
+skip_install = True
 commands =
-    sphinx-build -E -W -d docs/build/doctrees -b html docs docs/build/html
+    sphinx-build -T -E -W --keep-going {tty:--color} -d docs/build/doctrees -b html docs docs/build/html
```
Comment (webknjaz) [inline tox.ini]: `--keep-going` alone doesn't do anything unless you also add a nitpicky mode.
You can do
```suggestion
    sphinx-build -T -E -n -W --keep-going {tty:--color} -d docs/build/doctrees -b html docs docs/build/html
```
to match what you've enabled on RTD. Or alternatively, add `nitpicky = True` to `docs/conf.py` and it'll do the same.

Comment (webknjaz) [reply tox.ini]: Another option that may be useful to add is `-j auto` unless you have extensions that don't support it.

Comment (samdoran) [reply tox.ini]: The build is actually pretty quick right now. I don't know if we would benefit from parallelizing it.

Comment (webknjaz) [reply tox.ini]: It's just what people typically add by default so I thought I'd mention this.

Comment (samdoran) [reply tox.ini]: I had forgotten about this lame note:

> For security reasons, parallel mode is disabled on macOS and python3.8 and above

Comment (webknjaz) [reply tox.ini]: Ah right, that corner case is probably annoying for people who tolerate macs :)

Comment (samdoran) [reply tox.ini]: I fixed `sphinx` to use the proper start method. Doesn't seem to vastly improve the build time.

Single:
```
Executed in    5.50 secs    fish           external
   usr time    4.92 secs  107.00 micros    4.92 secs
   sys time    0.47 secs  957.00 micros    0.47 secs
```

Parallel:
```
Executed in    5.08 secs    fish           external
   usr time    7.63 secs    0.11 millis    7.63 secs
   sys time    1.45 secs    1.00 millis    1.45 secs
```

Comment (samdoran) [reply tox.ini]: I went on a side quest. https://github.com/sphinx-doc/sphinx/pull/9793

**Comment by samdoran** [comment]
@webknjaz Thanks for all your feedback. It's super helpful!

**Comment by webknjaz** [comment]
@samdoran just to clarify — did you explicitly decide not to use `-n` or did you miss that suggestion?

**Comment by samdoran** [comment]
@webknjaz I did miss that, but it's not in the RTD build command and `--keep-going` seemed to actually do the right thing without `-n` in my testing.

**Comment by samdoran** [comment]
Ok, I think I see the distinction. The warning I was seeing was due to a missing dependency in the virtual environment, not a missing doc reference:

```
WARNING: autodoc: failed to import module 'exceptions' from module 'ansible_runner'; the following exception was raised:
No module named 'pexpect'
```

**Comment by webknjaz** [comment]
> `--keep-going` seemed to actually do the right thing without `-n` in my testing.

To clarify: adding `-W` makes more warnings show up, they all are printed out in the output. `-n` on top of that makes Sphinx fail on the first warning it sees so you would only see one warning, fix it and have to rerun to see the next one. When you add `--keep-going` to the mix, it will not fail on the first warning but it will print out everything produced by `-W`, finish the build and only then return an error.

File: tox.ini [hunk 7]
```diff
@@ -27,5 +27,6 @@ commands=
 [testenv:docs]
 description = Build documentation
 deps = -r{toxinidir}/docs/requirements.txt
+skip_install = True
 commands =
-    sphinx-build -E -W -d docs/build/doctrees -b html docs docs/build/html
+    sphinx-build -T -E -W -n --keep-going {tty:--color} -j auto -d docs/build/doctrees -b html docs docs/build/html
```
Comment (webknjaz) [inline tox.ini]: Side question: why is the caching disabled with `-E`? Doesn't it conflict with the speed-up that `-d` would've caused?

**Review by webknjaz — APPROVED** [review]

File: docs/requirements.in (hunk 3 above)
Comment (samdoran) [reply docs/requirements.in]: I'm not sure about that one.

File: tox.ini (hunk 7 above)
Comment (samdoran) [reply tox.ini]: I'm not sure about that one.

**Comment by webknjaz** [comment]
Was `-n` causing problems locally?

**Comment by samdoran** [comment]
Yes. And the CI job failed too. Evidently we have some docs issues to resolve.

**Comment by webknjaz** [comment]
@samdoran you can add all reference errors to `nitpick_ignore` in `conf.py`. Then, you'd be able to keep `-n` and fix those over time.

**Comment by webknjaz** [comment]
Linking this for history: https://github.com/readthedocs/readthedocs.org/issues/8616
